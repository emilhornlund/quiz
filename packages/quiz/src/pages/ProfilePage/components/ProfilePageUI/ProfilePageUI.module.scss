@use '../../../../styles/helpers';
@use '../../../../styles/variables';

.profileDetailsForm {
  display: flex;
  flex-direction: row;

  @include helpers.mobile {
    column-gap: variables.$mobile-narrow-layout-gutter;
    width: 100%;
  }

  @include helpers.tablet {
    column-gap: variables.$tablet-narrow-layout-gutter;
    width: variables.$mobile-width;
  }

  @include helpers.desktop {
    column-gap: variables.$desktop-narrow-layout-gutter;
    width: variables.$mobile-width;
  }

  .column {
    display: flex;
    flex-direction: column;

    &.full {
      flex: 1;
    }
  }
}
