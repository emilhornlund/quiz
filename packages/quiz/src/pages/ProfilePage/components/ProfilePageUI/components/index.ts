export type { ProfileDetailsProps } from './ProfileDetails'
export type { ProfileQuizzesProps } from './ProfileQuizzes'
export { default as ProfileDetails } from './ProfileDetails'
export { default as ProfileQuizzes } from './ProfileQuizzes'
