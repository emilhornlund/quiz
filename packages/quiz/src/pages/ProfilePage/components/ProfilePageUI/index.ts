export type { ProfilePageUIProps } from './ProfilePageUI'
export { default } from './ProfilePageUI'
