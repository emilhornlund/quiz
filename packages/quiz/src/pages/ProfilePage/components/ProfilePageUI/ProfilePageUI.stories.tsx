import type { Meta, StoryObj } from '@storybook/react'
import { withRouter } from 'storybook-addon-remix-react-router'
import { v4 as uuidv4 } from 'uuid'

import { Player } from '../../../../models'

import ProfilePageUI from './ProfilePageUI'

const meta = {
  title: 'Pages/ProfilePage',
  component: ProfilePageUI,
  decorators: [withRouter],
  parameters: {
    layout: 'fullscreen',
  },
} satisfies Meta<typeof ProfilePageUI>

export default meta
type Story = StoryObj<typeof meta>

const player: Player = {
  id: uuidv4(),
  nickname: 'FrostyBear',
}

export const Default = {
  args: {
    player,
    onNicknameChange: () => undefined,
  },
} satisfies Story
