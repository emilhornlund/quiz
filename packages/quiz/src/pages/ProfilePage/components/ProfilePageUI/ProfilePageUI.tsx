import { faFloppyDisk } from '@fortawesome/free-solid-svg-icons'
import {
  EMAIL_MAX_LENGTH,
  EMAIL_MIN_LENGTH,
  EMAIL_REGEX,
  FAMILY_NAME_MAX_LENGTH,
  FAMILY_NAME_MIN_LENGTH,
  FAMILY_NAME_REGEX,
  GIVEN_NAME_MAX_LENGTH,
  GIVEN_NAME_MIN_LENGTH,
  GIVEN_NAME_REGEX,
  PLAYER_NICKNAME_MAX_LENGTH,
  PLAYER_NICKNAME_MIN_LENGTH,
  PLAYER_NICKNAME_REGEX,
  UpdateUserProfileRequestDto,
} from '@quiz/common'
import React, { FC, FormEvent, useEffect, useMemo, useState } from 'react'

import { Button, Page, TextField, Typography } from '../../../../components'

import styles from './ProfilePageUI.module.scss'

export type UpdateUserProfileFormFields = UpdateUserProfileRequestDto

export interface ProfilePageUIProps {
  values: UpdateUserProfileFormFields
  loading: boolean
  onChange: (request: UpdateUserProfileFormFields) => void
}

const ProfilePageUI: FC<ProfilePageUIProps> = ({
  values,
  loading,
  onChange,
}) => {
  const [formFields, setFormFields] =
    useState<UpdateUserProfileFormFields>(values)

  useEffect(() => {
    setFormFields(values)
  }, [values])

  const handleChangeFormField = <K extends keyof UpdateUserProfileFormFields>(
    key: K,
    value: UpdateUserProfileFormFields[K],
  ) => {
    setFormFields({ ...formFields, [key]: value })
  }

  const [validFormFields, setValidFormFields] = useState<{
    [key in keyof UpdateUserProfileFormFields]: boolean
  }>({
    email: false,
    givenName: false,
    familyName: false,
    defaultNickname: false,
  })

  const isFormValid = useMemo(
    () => Object.values(validFormFields).every((valid) => !!valid),
    [validFormFields],
  )

  const handleChangeValidFormField = <
    K extends keyof UpdateUserProfileFormFields,
  >(
    key: K,
    valid: boolean,
  ) => {
    if (validFormFields[key] !== valid) {
      setValidFormFields((prevState) => {
        return { ...prevState, [key]: valid }
      })
    }
  }

  const handleSubmit = (event: FormEvent) => {
    event.preventDefault()
    onChange(formFields)
  }

  return (
    <Page align="start" discover profile>
      <Typography variant="subtitle">Shape your quiz identity</Typography>
      <Typography variant="text" size="medium">
        Update your personal details to keep your profile up to date. Your
        information helps personalize your quiz experience and lets others
        recognize you during games.
      </Typography>
      <form className={styles.profileDetailsForm} onSubmit={handleSubmit}>
        <TextField
          id="email"
          type="text"
          placeholder="Email"
          value={formFields.email}
          minLength={EMAIL_MIN_LENGTH}
          maxLength={EMAIL_MAX_LENGTH}
          regex={EMAIL_REGEX}
          disabled={loading}
          onChange={(value) => handleChangeFormField('email', value as string)}
          onValid={(valid) => handleChangeValidFormField('email', valid)}
          required
        />
        <TextField
          id="givenName"
          type="text"
          placeholder="Given Name"
          value={formFields.givenName}
          minLength={GIVEN_NAME_MIN_LENGTH}
          maxLength={GIVEN_NAME_MAX_LENGTH}
          regex={GIVEN_NAME_REGEX}
          disabled={loading}
          onChange={(value) =>
            handleChangeFormField('givenName', value as string)
          }
          onValid={(valid) => handleChangeValidFormField('givenName', valid)}
        />
        <TextField
          id="familyName"
          type="text"
          placeholder="Family Name"
          value={formFields.familyName}
          minLength={FAMILY_NAME_MIN_LENGTH}
          maxLength={FAMILY_NAME_MAX_LENGTH}
          regex={FAMILY_NAME_REGEX}
          disabled={loading}
          onChange={(value) =>
            handleChangeFormField('familyName', value as string)
          }
          onValid={(valid) => handleChangeValidFormField('familyName', valid)}
        />
        <TextField
          id="defaultNickname"
          type="text"
          placeholder="Default Nickname"
          value={formFields.defaultNickname}
          minLength={PLAYER_NICKNAME_MIN_LENGTH}
          maxLength={PLAYER_NICKNAME_MAX_LENGTH}
          regex={PLAYER_NICKNAME_REGEX}
          disabled={loading}
          onChange={(value) =>
            handleChangeFormField('defaultNickname', value as string)
          }
          onValid={(valid) =>
            handleChangeValidFormField('defaultNickname', valid)
          }
        />
        <Button
          id="update-user-button"
          type="submit"
          kind="call-to-action"
          size="normal"
          value="Save"
          icon={faFloppyDisk}
          loading={loading}
          disabled={!isFormValid || loading}
        />
      </form>
    </Page>
  )
}

export default ProfilePageUI
