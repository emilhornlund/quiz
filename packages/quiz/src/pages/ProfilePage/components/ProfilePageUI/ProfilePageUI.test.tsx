import { render } from '@testing-library/react'
import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { describe, expect, it } from 'vitest'

import ProfilePageUI from './ProfilePageUI'

describe('ProfilePageUI', () => {
  it('should render ProfilePageUI', async () => {
    const { container } = render(
      <MemoryRouter>
        <ProfilePageUI
          values={{
            email: '',
            givenName: '',
            familyName: '',
            defaultNickname: '',
          }}
          loading={false}
          onChange={() => undefined}
        />
      </MemoryRouter>,
    )

    expect(container).toMatchSnapshot()
  })
})
