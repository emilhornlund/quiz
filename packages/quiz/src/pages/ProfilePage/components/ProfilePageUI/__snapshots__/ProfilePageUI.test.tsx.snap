// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfilePageUI > should render ProfilePageUI 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Quiz
        </span>
      </button>
      <div
        class="side"
      >
        <a
          href="/discover"
        >
          Discover
        </a>
        <div
          class="menuButtonWrapper"
        >
          <button
            type="button"
          >
            <img
              alt="Profile"
              src="/src/assets/images/avatar.svg"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="content fullWidth startAlign"
    >
      <div
        class="typography subtitle full"
      >
        Update Your Profile
      </div>
      <div
        class="typography text medium"
      >
        Customize your player profile by updating your nickname. Your nickname is how other players will see you in quizzes.
      </div>
      <form
        class="profileDetailsForm"
      >
        <div
          class="column full"
        >
          <div
            class="inputContainer"
          >
            <div
              class="textFieldInputContainer textFieldInputKindPrimary"
            >
              <input
                class="textfield"
                data-testid="test-nickname-textfield"
                id="nickname"
                maxlength="20"
                minlength="2"
                name="nickname"
                placeholder="Nickname"
                type="text"
                value="FrostyBear"
              />
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="buttonInputContainer buttonInputKindCallToAction"
          >
            <button
              data-testid="test-update-player-button-button"
              id="update-player-button"
              name="update-player-button"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-floppy-disk "
                data-icon="floppy-disk"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M64 32C28.7 32 0 60.7 0 96L0 416c0 35.3 28.7 64 64 64l320 0c35.3 0 64-28.7 64-64l0-242.7c0-17-6.7-33.3-18.7-45.3L352 50.7C340 38.7 323.7 32 306.7 32L64 32zm0 96c0-17.7 14.3-32 32-32l192 0c17.7 0 32 14.3 32 32l0 64c0 17.7-14.3 32-32 32L96 224c-17.7 0-32-14.3-32-32l0-64zM224 288a64 64 0 1 1 0 128 64 64 0 1 1 0-128z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
